### PR TITLE
feat(index): add `options` validation (`schema-utils`)

### LINKED ISSUE #4: feat: add options validation
It should be processed through [ajv](https://www.npmjs.com/package/ajv) to give proper errors if the schema validation fails.

### PR DESCRIPTION
**Pull request by hulkish**
Closes #4, https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/64.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -150,11 +140,11 @@ class UglifyJsPlugin {
...
     const requestShortener = new RequestShortener(compiler.context);
-    // Copy uglify options
-    const uglifyOptions = UglifyJsPlugin.buildDefaultUglifyOptions(this.uglifyOptions);
-    // Making sure output options exists if there is an extractComments options
-    if (this.options.extractComments) {
-      uglifyOptions.output = uglifyOptions.output || {};
+    const { uglifyOptions } = this.options;
+    const uglifiedAssets = new WeakSet();
+
+    if (typeof this.options.sourceMap === 'undefined' && (compiler.options.devtool === 'sourcemap' || compiler.options.devtool === 'source-map')) {
```
Comment (evilebottnawi) [inline src/index.js]: Seems not related to this PR

Comment (hulkish) [inline src/index.js]: You're right. I can separate this into a diff pr if u like. 

However, I gave this a good amount of thought and I believe it's been a missing piece to this plugin for some time now.

Not to mention - [this same logic is built into webpack core](https://github.com/webpack/webpack/blob/0ab36c491ea547add35af0c3fbbcdfab03a4c15e/bin/convert-argv.js#L441). I believe this responsibility belongs in this plugin.

If this pr is merged i plan to follow up by making a pr at webpack core repo to extract this logic from there.

Comment (evilebottnawi) [inline src/index.js]: @hulkish let's wait what other will say, but be good have separate PR for this, it is allow to maintenance CHANGELOG

Comment (hulkish) [inline src/index.js]: ok sure np

Comment (evilebottnawi) [inline src/index.js]: > If this pr is merged i plan to follow up by making a pr at webpack core repo to extract this logic from there. This also doesn't introduce a breaking change and all tests are passing.

btw good enchantment!

Comment (hulkish) [inline src/index.js]: bahaha : 🌮

File: src/index.js [hunk 2]
```diff
@@ -7,49 +7,39 @@ import { SourceMapConsumer } from 'source-map';
...
 /* eslint-disable
   no-param-reassign
 */
 
 const warningRegex = /\[.+:([0-9]+),([0-9]+)\]/;
 
-const defaultUglifyOptions = {
-  output: {
-    comments: /^\**!|@preserve|@license|@cc_on/,
-    beautify: false,
```
Comment (hulkish) [inline src/index.js]: Removed these because these are the same defaults `uglify-es` uses internallu.

File: package.json [hunk 3]
```diff
@@ -10,7 +10,7 @@
   ],
   "scripts": {
     "start": "npm run build -- -w",
-    "build": "cross-env NODE_ENV=production babel src -d dist --ignore 'src/**/*.test.js'",
```
Comment (hulkish) [inline package.json]: added this because babel-cli to auto-copy src/options-schema.json to the dist folder. (to prevent MODULE_NOT_FOUND error on published code)

File: src/index.js [hunk 4]
```diff
@@ -7,49 +7,39 @@ import { SourceMapConsumer } from 'source-map';
...
+      test = /\.js$/i,
+      warningsFilter = () => true,
+      extractComments,
+      sourceMap,
+    } = options;
 
-    this.uglifyOptions = this.options.uglifyOptions || {};
-  }
-
-  static buildDefaultUglifyOptions({ ecma, warnings, parse = {}, compress = {}, mangle, output, toplevel, ie8 }) {
```
Comment (hulkish) [inline src/index.js]: Simplified this by leveraging destructing default values.

File: src/index.js [hunk 5]
```diff
@@ -166,7 +156,6 @@ class UglifyJsPlugin {
       }
 
       compilation.plugin('optimize-chunk-assets', (chunks, callback) => {
-        const uglifiedAssets = new WeakSet();
```
Comment (hulkish) [inline src/index.js]: Hoisted this to `apply` scope. This was my change originally and was an optimization that I wanted to include but was late to do so before it got merged. I've tested it pretty thoroughly.

File: src/options-schema.json [hunk 6]
```diff
@@ -0,0 +1,50 @@
+{
```
Comment (hulkish) [inline src/options-schema.json]: Reworked from #64

File: src/index.js [hunk 7]
```diff
@@ -150,11 +140,11 @@ class UglifyJsPlugin {
 
   apply(compiler) {
     const requestShortener = new RequestShortener(compiler.context);
-    // Copy uglify options
-    const uglifyOptions = UglifyJsPlugin.buildDefaultUglifyOptions(this.uglifyOptions);
-    // Making sure output options exists if there is an extractComments options
-    if (this.options.extractComments) {
-      uglifyOptions.output = uglifyOptions.output || {};
```
Comment (hulkish) [inline src/index.js]: replaced with logic in constructor

File: src/index.js [hunk 8]
```diff
@@ -7,49 +7,39 @@ import { SourceMapConsumer } from 'source-map';
 import { SourceMapSource, RawSource, ConcatSource } from 'webpack-sources';
 import RequestShortener from 'webpack/lib/RequestShortener';
 import ModuleFilenameHelpers from 'webpack/lib/ModuleFilenameHelpers';
+import validateOptions from 'schema-utils';
 import uglify from 'uglify-es';
+import optionsSchema from './options-schema.json';
```
Comment (michael-ciniawsky) [inline src/index.js]: `optionsSchema` => `schema`

File: src/index.js [hunk 9]
```diff
@@ -7,49 +7,39 @@ import { SourceMapConsumer } from 'source-map';
...
-
 class UglifyJsPlugin {
-  constructor(options) {
-    if (typeof options !== 'object' || Array.isArray(options)) {
-      this.options = {};
-    } else {
-      this.options = options || {};
-    }
+  constructor(options = {}) {
+    validateOptions(optionsSchema, options, 'UglifyJsPlugin');
```
Comment (michael-ciniawsky) [inline src/index.js]: `UglifyJsPlugin` => `UglifyJS Plugin`

File: src/options-schema.json (hunk 6 above)
Comment (michael-ciniawsky) [inline src/options-schema.json]: `src/options-schema.json` => `./options.json`, if  it should reside in the src and needs to be copied, the change needs to be made in  `webpack-defaults` beforehand as always 😛

File: src/options-schema.json [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        "mangle": {
+          "type": ["boolean", "object"],
+          "properties": {
+            "properties": {
+              "type": ["boolean", "object"]
+            }
+          }
+        },
+        "output": {
+          "type": ["object", "null"]
```
Comment (michael-ciniawsky) [inline src/options-schema.json]: Is `null` valid ? 🙃

File: src/options-schema.json [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    "sourceMap": {
+      "type": "boolean"
+    },
+    "uglifyOptions": {
+      "additionalProperties": true,
+      "type": "object",
+      "properties": {
+        "ecma": {
+          "type": "number",
+          "enum": [5, 6, 7, 8]
```
Comment (michael-ciniawsky) [inline src/options-schema.json]: Could make this throw and see if the `ValidationError` is still displayed correctly with this additional info please?

File: package.json (hunk 3 above)
Comment (michael-ciniawsky) [inline package.json]: => `webpack-defaults`, but `options.json` in the root is also ok imho,  since it is metadata, still needs to be added to `pkg.files` then, so really no preference :)

File: src/options-schema.json (hunk 10 above)
Comment (hulkish) [inline src/options-schema.json]: yes, it's default value is null: https://github.com/mishoo/UglifyJS2#user-content-minify-options

Comment (michael-ciniawsky) [inline src/options-schema.json]: I was referring to the schema validation here :) e.g `{Regex}, {Function}` aren't supported  by `ajv` (`schema-utils`) yet.

**Comment by hulkish** [comment]
ok done @michael-ciniawsky https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/80/commits/2a3687e9de2fca2d22cd0366226da9b893020d8e

File: options.json [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    "test": {},
+    "warningsFilter": {},
+    "extractComments": {},
+    "sourceMap": {
+      "type": "boolean"
+    },
+    "uglifyOptions": {
+      "additionalProperties": true,
+      "type": "object",
+      "properties": {
```
Comment (michael-ciniawsky) [inline options.json]: `ie8` ?

File: options.json [hunk 13]
```diff
@@ -0,0 +1,51 @@
+{
+  "type": "object",
+  "properties": {
```
Comment (michael-ciniawsky) [inline options.json]: `include, exclude, parallel` ?

File: package.json [hunk 14]
```diff
@@ -6,7 +6,8 @@
   "license": "MIT",
   "main": "dist/cjs.js",
   "files": [
-    "dist"
+    "dist",
+    "options.json"
```
Comment (michael-ciniawsky) [inline package.json]: I'm 💯 sry about this, your initial way of handling this file via `--copy-files` in the `build` script is the better solution, please revert :) 👍

File: src/index.js [hunk 15]
```diff
@@ -7,6 +7,8 @@ import { SourceMapConsumer } from 'source-map';
 import { SourceMapSource, RawSource, ConcatSource } from 'webpack-sources';
 import RequestShortener from 'webpack/lib/RequestShortener';
 import ModuleFilenameHelpers from 'webpack/lib/ModuleFilenameHelpers';
+import validateOptions from 'schema-utils';
+import schema from '../options.json';
```
Comment (michael-ciniawsky) [inline src/index.js]: `../options.json` => `./options.json` (`../` => `./`) ?

Comment (hulkish) [inline src/index.js]: yea, I thought that you wanted me to include that with the --copy-files change.. ill switch it back
